src/WebViewShared.tsx

```diff
@@ -43,16 +43,54 @@ const _passesWhitelist = (
 ) => {
   try {
     const { origin } = new URL(url)
-    return origin && matchWithRegexList(compiledWhitelist, origin)
+    return Boolean(origin && matchWithRegexList(compiledWhitelist, origin))
   } catch {
     return false
   }
 };
 
 const compileWhitelist = (
   originWhitelist: readonly string[],
+  defaultEntries: readonly string[] = ['about:blank'],
 ): readonly RegExp[] =>
-  ['about:blank', ...(originWhitelist || [])].map(stringWhitelistToRegex);
+  [...defaultEntries, ...(originWhitelist || [])].map(stringWhitelistToRegex);
+
+const isDownloadMessageAllowed = ({
+  data,
+  url,
+  downloadWhitelist,
+}: {
+  data: string;
+  url: string;
+  downloadWhitelist: { origin: string; allowedFileExtensions: string[] }[];
+}): boolean => {
+  let parsedData;
+  
+  try {
+    parsedData = JSON.parse(data);
+  } catch {
+    return true; // Invalid JSON — treat as non-download message
+  }
+
+  if (parsedData.method !== 'download') {
+    return true;
+  }
+
+  const { origin } = new URL(url);
+  const fileExtension = parsedData.params?.fileName?.split('.').pop()?.toLowerCase();
+  
+  if (!fileExtension || !origin) {
+    return false;
+  }
+
+  const matchingRule = downloadWhitelist.find(rule => {
+    const ruleOriginRegex = stringWhitelistToRegex(rule.origin);
+
+    return ruleOriginRegex.test(origin) && rule.allowedFileExtensions.includes(fileExtension);
```

Comment:
Subject to 'type confusion'-like issue [^0][^1]. Type-check `allowedFileExtensions`:
```js
const fileExtension = 'js'

const downloadWhitelist = [{
  origin: 'https://example.com',
  allowedFileExtensions: 'json'
}]

const result = downloadWhitelist.find(rule => {
  return rule.allowedFileExtensions.includes(fileExtension)
})

!!result // oops
```

[^0]: https://cwe.mitre.org/data/definitions/843.html
[^1]: https://snyk.io/es/blog/remediate-javascript-type-confusion-bypassed-input-validation/